### PR TITLE
Spike/context menu groups

### LINKED ISSUE #499: Context Menu Stacks by Content Layers with Basic Groupings
## Description
As a Content Editor, I want to easily associate available content actions by area in which I am working in, so that I can quickly access the functions I need to work with. 
- Referring to when the Context Menu appears; The context menu layering

https://jhdfdc.axshare.com/#g=1&p=aries&id=3ct389 

## Acceptance Criteria
AC1:

The buttons of the context menu are divided into named sections

AC2:

The section titles are the Component Name, or Container (for the flow container) 

AC3. 

The layers are ordered horizontally with inner most (on top/to the left) to outer most based on the component they are applied to.

### PR DESCRIPTION
**Pull request by wodenx**
## Changes

- Provides customizable groups for context menu items.

## Related Issues
- Closes #499

### CODE REVIEW

File: packages/bodiless-components/src/List/withListButtons.tsx [hunk 1]
```diff
@@ -15,45 +15,68 @@
...
   // The first child list is the one to which this toggle applies,
   // so we check to see if more than one.
-  return descendants.filter(c => c.type === 'list-item').length > 1;
+  return descendants.filter(c => c.type === 'list-item').length > count;
 };
 
-const useMenuOptions = (props: ItemProps) => {
-  const context = useEditContext();
+const useMenuOptions = (useOverrides: UseListOverrides = () => ({})) => (props: ItemProps) => {
+  // const context = useEditContext();
```
Comment (beliayeu) [inline packages/bodiless-components/src/List/withListButtons.tsx:34]: remove?

File: packages/bodiless-core/__tests__/PageEditContext.test.ts [hunk 2]
```diff
@@ -322,22 +324,6 @@ describe('Update menu options', () => {
     expect(optionListener).toHaveBeenCalledTimes(0);
   });
 
-  it('Does not notify when new context with same menu options activates', () => {
```
Comment (beliayeu) [inline packages/bodiless-core/__tests__/PageEditContext.test.ts:325]: are these tests deleted on purpose?

**Review by beliayeu — APPROVED** [review]

File: packages/bodiless-core/__tests__/PageEditContext.test.ts (hunk 2 above)
Comment (wodenx) [reply packages/bodiless-core/__tests__/PageEditContext.test.ts:325]: yep. can't remember why exactly, but some change made them no longer relevant...

File: packages/bodiless-components/src/List/asChameleonSubList.tsx [hunk 3]
```diff
@@ -1,58 +1,55 @@
...
-  const context = useEditContext();
   return {
-    isHidden: useCallback(() => isOn || hasChildSubList(context), [isOn]),
+    isHidden: isOn,
     icon: 'playlist_add',
-    label: 'Sub',
     name: `chameleon-sublist-${v1()}`,
+    label: 'Sub',
+    groupMerge: 'merge-up',
+    // label: 'Add',
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-components/src/List/asChameleonSubList.tsx:31]: fixed in 1abc3f28f5b3500eb4978fb1e9fc75abe3c2b1aa

File: packages/bodiless-core/__tests__/withEditButton.test.tsx [hunk 4]
```diff
@@ -21,6 +21,7 @@ import { observer } from 'mobx-react-lite';
 import withEditButton from '../src/withEditButton';
 import { useEditContext } from '../src/hooks';
 import ContextMenuItem from '../src/components/ContextMenuItem';
+import { TMenuOption, EditButtonOptions } from '../lib';
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-core/__tests__/withEditButton.test.tsx]: fixed in 1abc3f28f5b3500eb4978fb1e9fc75abe3c2b1aa

File: packages/bodiless-core/__tests__/withEditButton.test.tsx [hunk 5]
```diff
@@ -148,22 +149,33 @@ describe('withEditButton', () => {
...
     );
-    expect(wrapper.prop('name')).toBe(options.name);
-    const menuOptions = wrapper.prop('getMenuOptions')();
-    expect(menuOptions.length).toBe(1);
-    expect(menuOptions[0].icon).toBe(options.icon);
-    expect(menuOptions[0].name).toBe(options.name);
-    expect(menuOptions[0].global).toBe(false);
-    expect(menuOptions[0].local).toBeUndefined();
+
+    console.log(wrapper.debug());
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-core/__tests__/withEditButton.test.tsx]: fixed in 1abc3f28f5b3500eb4978fb1e9fc75abe3c2b1aa

File: packages/bodiless-core/src/ContextMenu/StructuredChildren.tsx [hunk 6]
```diff
@@ -16,7 +16,7 @@ import React, {
   FC, ReactElement, ReactNode, ComponentType,
 } from 'react';
 import { flow } from 'lodash';
-import { ContextMenuGroupProps } from '../Types/ContextMenuTypes';
+import { IContextMenuItemProps } from '../Types/ContextMenuTypes';
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-core/src/ContextMenu/StructuredChildren.tsx]: fixed in 1abc3f28f5b3500eb4978fb1e9fc75abe3c2b1aa

File: packages/bodiless-core/src/PageEditContext/Store.tsx [hunk 7]
```diff
@@ -122,6 +122,7 @@ export class PageEditStore implements PageEditStoreInterface {
           Object.assign(existing, next);
           extendObservable(existing, newProps);
         } else {
+          // map!.set(op.name, { ...next, context: c });
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-core/src/PageEditContext/Store.tsx]: fixed in 1abc3f28f5b3500eb4978fb1e9fc75abe3c2b1aa

File: packages/bodiless-components/src/List/types.ts [hunk 8]
```diff
@@ -14,7 +14,7 @@
 
 import { HTMLProps, ComponentType, PropsWithChildren } from 'react';
 import { DesignableComponentsProps, DesignableProps } from '@bodiless/fclasses';
-import { WithNodeProps } from '@bodiless/core';
+import { WithNodeProps, EditButtonOptions } from '@bodiless/core';
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-components/src/List/types.ts]: fixed in 1abc3f28f5b3500eb4978fb1e9fc75abe3c2b1aa

File: packages/bodiless-components/src/List/withListButtons.tsx [hunk 9]
```diff
@@ -15,45 +15,68 @@
...
+  const { groupLabel = defaultGroupLabel, global = false, local = true } = useOverrides(props);
+  const id = v1();
+  const group = `list-item-group-${id}`;
+
+  const menuOptions:TMenuOption[] = useMemo(() => ([
     {
-      name: `add-${v1()}`,
-      isHidden: () => hasChildList(context),
+      name: `add-${id}`,
+      // isHidden: () => hasChildSubList(context),
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-components/src/List/withListButtons.tsx]: fixed in 1abc3f28f5b3500eb4978fb1e9fc75abe3c2b1aa

File: packages/bodiless-components/src/List/withListButtons.tsx (hunk 1 above)
Comment (wodenx) [reply packages/bodiless-components/src/List/withListButtons.tsx]: fixed in 1abc3f28f5b3500eb4978fb1e9fc75abe3c2b1aa

**Review by beliayeu — APPROVED** [review]

**Review by beliayeu — APPROVED** [review]

**Review by hvanyo — APPROVED** [review]

**Pull request merged**